### PR TITLE
Block Library: Allow changing Site Title block heading level.

### LINKED ISSUE #20333: Site Title block: do not force h1 on all pages
**Is your feature request related to a problem? Please describe.**
The `site-title` block currently uses an `h1` element at all times:

https://github.com/WordPress/gutenberg/blob/32aaed94dcc5c9b27889a1819142d4de1d6f7e94/packages/block-library/src/site-title/index.php#L14

Subpages and archives will wrap the site title in an h1 as a result. This is less good for accessibility and SEO, because the site title is not normally the most important title on those pages — the page/post/archive title is.

Popular themes typically use `h1` only if `( is_front_page() || is_home() ) && ! is_page()`, and a `div` or `p` at other times.

From [Twenty Twenty](https://github.com/WordPress/twentytwenty/blob/6d0a5240af108d02f58ec797fd49b32458a4c698/inc/template-tags.php#L43-L45
):

```php
		'home_wrap'   => '<h1 class="%1$s">%2$s</h1>',
		'single_wrap' => '<div class="%1$s faux-heading">%2$s</div>',
		'condition'   => ( is_front_page() || is_home() ) && ! is_page(),
```

Genesis does something similar, but lets users choose whether to use h1 or p on the homepage, defaulting to h1 for “latest posts” and p for static pages.

**Describe the solution you'd like**
Should WP behave like Twenty Twenty, Genesis, and others instead of forcing an h1 everywhere and placing the burden on theme developers to adjust it?

Users could still filter the result if they wish, but it would be more optimal by default.

**Describe alternatives you've considered**
Users who care about accessibility and SEO will need to filter `site-title` block output or use a theme or SEO plugin that handles this for them.

### PR DESCRIPTION
**Pull request by epiqueras**
Closes #20333

## Description

This PR makes the heading level of the Site Title block configurable through the block toolbar. It allows you to set it to all six heading levels or a paragraph level.

It also fixes a bug in `core-data` where a null value was having properties accessed.

## How has this been tested?

It was verified that changing the heading level works as expected in the editor and the front end.

## Types of Changes

*New Feature*: The Site Title block's heading level can now be configured.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by johnstonphilip** [comment]
@epiqueras This is great!

Screen Shot 2020-02-21 at 1 38 02 PM

I am curious if it makes sense to add a `None` option so that it can be used inline. Say, as part of a blog post body as a nested block:

Example:
`Here's another blog post coming at you from [site_title_block_here]. We love to change our name often though, so keep an eye on that sentence for when we change it again next week!`

Thoughts?

**Comment by johnstonphilip** [comment]
One issue I just noticed as well, is that while changing the setting in wp-admin works, it does not change the output on the frontend, as it still uses `<h1>`.

**Comment by epiqueras** [comment]
>I am curious if it makes sense to add a None option so that it can be used inline. Say, as part of a blog post body as a nested block:

I am not sure; I think this block needs a proper full design with more features like colors and alignments.

>One issue I just noticed as well, is that while changing the setting in wp-admin works, it does not change the output on the frontend, as it still uses `<h1>`.

Are you sure you rebuilt things? I see it changing as expected.

**Comment by johnstonphilip** [comment]
@epiqueras I did double check that yeah. Everything has rebuilt properly, and I am seeing the js changes from this PR in action.

I think the issue might be that I have the Site Title block in both my header.html file, and in the body of a post on the same page. 

I added some debugging to the output function, and the first time the attributes are an empty array, and the second time it has level 3, though both are still using `H1`.

Screen Shot 2020-02-21 at 2 03 10 PM

**Comment by shaunandrews** [comment]
Its kind of weird to have the Paragraph option in the list. I expected that I'd get all the features of the normal Paragraph block (bolding and the like), but it doesn't do that.

**Comment by epiqueras** [comment]
@johnstonphilip Short-circuiting wasn't working right; it's fixed now.

>Its kind of weird to have the Paragraph option in the list. I expected that I'd get all the features of the normal Paragraph block (bolding and the like), but it doesn't do that.

That would be a block transformation. Where would you put it instead?

**Comment by johnstonphilip** [comment]
@epiqueras Working now! 👍

**Comment by epiqueras** [comment]
Thanks for testing!

**Comment by johnstonphilip** [comment]
@epiqueras I think you mentioned this but I just want to confirm: Based on my testing, this also appears to fix https://github.com/WordPress/gutenberg/issues/20337

Is that right?

**Comment by epiqueras** [comment]
Yes, it does.

**Comment by shaunandrews** [comment]
> That would be a block transformation. Where would you put it instead?

Sorry, I'm not suggesting a transformation.

I think my hesitation is that this menu is essentially changing the `html` tag that is wrapped around the site title. I haven't seen an example where we do this in the toolbar, nor has I seen an instance where we've combined both heading levels with the concept of a "paragraph level" option. This blurs the line between the font-size, and the semantic tag used.

That said, if we want to keep the paragraph option, I think it should be the last in the list. Its both visually the smallest, and semantically less important.

Site Title Size

**Comment by paaljoachim** [comment]
Is there any connection with this PR?
https://github.com/WordPress/gutenberg/pull/20246

**Comment by epiqueras** [comment]
No

File: packages/block-library/src/site-title/edit/level-icon.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+		4: 'M9 15H7v-4H3v4H1V5h2v4h4V5h2v10zm10-2h-1v2h-2v-2h-5v-2l4-6h3v6h1v2zm-3-2V7l-2.8 4H16z',
+		5: 'M12.1 12.2c.4.3.7.5 1.1.7.4.2.9.3 1.3.3.5 0 1-.1 1.4-.4.4-.3.6-.7.6-1.1 0-.4-.2-.9-.6-1.1-.4-.3-.9-.4-1.4-.4H14c-.1 0-.3 0-.4.1l-.4.1-.5.2-1-.6.3-5h6.4v1.9h-4.3L14 8.8c.2-.1.5-.1.7-.2.2 0 .5-.1.7-.1.5 0 .9.1 1.4.2.4.1.8.3 1.1.6.3.2.6.6.8.9.2.4.3.9.3 1.4 0 .5-.1 1-.3 1.4-.2.4-.5.8-.9 1.1-.4.3-.8.5-1.3.7-.5.2-1 .3-1.5.3-.8 0-1.6-.1-2.3-.4-.6-.2-1.1-.6-1.6-1-.1-.1 1-1.5 1-1.5zM9 15H7v-4H3v4H1V5h2v4h4V5h2v10z',
+		6: 'M9 15H7v-4H3v4H1V5h2v4h4V5h2v10zm8.6-7.5c-.2-.2-.5-.4-.8-.5-.6-.2-1.3-.2-1.9 0-.3.1-.6.3-.8.5l-.6.9c-.2.5-.2.9-.2 1.4.4-.3.8-.6 1.2-.8.4-.2.8-.3 1.3-.3.4 0 .8 0 1.2.2.4.1.7.3 1 .6.3.3.5.6.7.9.2.4.3.8.3 1.3s-.1.9-.3 1.4c-.2.4-.5.7-.8 1-.4.3-.8.5-1.2.6-1 .3-2 .3-3 0-.5-.2-1-.5-1.4-.9-.4-.4-.8-.9-1-1.5-.2-.6-.3-1.3-.3-2.1s.1-1.6.4-2.3c.2-.6.6-1.2 1-1.6.4-.4.9-.7 1.4-.9.6-.3 1.1-.4 1.7-.4.7 0 1.4.1 2 .3.5.2 1 .5 1.4.8 0 .1-1.3 1.4-1.3 1.4zm-2.4 5.8c.2 0 .4 0 .6-.1.2 0 .4-.1.5-.2.1-.1.3-.3.4-.5.1-.2.1-.5.1-.7 0-.4-.1-.8-.4-1.1-.3-.2-.7-.3-1.1-.3-.3 0-.7.1-1 .2-.4.2-.7.4-1 .7 0 .3.1.7.3 1 .1.2.3.4.4.6.2.1.3.3.5.3.2.1.5.2.7.1z',
+	};
```
Comment (ockham) [inline packages/block-library/src/site-title/edit/level-icon.js]: Would we save memory if we moved this const out of the function (and into module level) so it's not re-initialized each time the function is called?

Comment (epiqueras) [reply packages/block-library/src/site-title/edit/level-icon.js]: Probably not a relevant amount.

**Review by ockham — APPROVED** [review]
Looking good overall!

I [agree](https://github.com/WordPress/gutenberg/pull/20361#issuecomment-589828015) with Shaun that if we're going to keep the option to format the site title as a paragraph, the paragraph option should be last in the format picker (even if that makes looping over 0..6 harder).

Another (more architectural yet vague) question: I seem to recall that there were plans to support composing blocks from others, with the composed (outermost) block controlling the composing blocks' attributes, and having the composed block 'inherit' the toolbar (and inspector?) controls from the composing ones. (Unfortunately, I can't seem to find any references right now.) Would it thus be possible to compose the site title block from a heading block?

**Comment by epiqueras** [comment]
>Would it thus be possible to compose the site title block from a heading block?

That's more for blocks that have a hierarchy, e.g., nav menu -> nav link.

**Comment by ockham** [comment]
> > Would it thus be possible to compose the site title block from a heading block?
> 
> That's more for blocks that have a hierarchy, e.g., nav menu -> nav link.

Not sure we're on the same page here, I don't mean the established Inner Blocks pattern, but a pattern that I thought has been emerging more recently -- with a more rigid connection between the 'outer' (composed) and 'inner' (composing) blocks. I thought I'd seen something to that effect, but maybe I'm mistaken :thinking:

**Comment by ockham** [comment]
> > > Would it thus be possible to compose the site title block from a heading block?
> > 
> > 
> > That's more for blocks that have a hierarchy, e.g., nav menu -> nav link.
> 
> Not sure we're on the same page here, I don't mean the established Inner Blocks pattern, but a pattern that I thought has been emerging more recently -- with a more rigid connection between the 'outer' (composed) and 'inner' (composing) blocks. I thought I'd seen something to that effect, but maybe I'm mistaken

Ahh, I think I was mostly thinking of your own context concepts from #19685 and #19572 :blush:

**Comment by epiqueras** [comment]
That still is for hierarchical blocks.

The only block I can think of that absorbs the toolbar, as you describe, is the Navigation block.

**Pull request merged**